python/whylogs/extras/nlp_metric.py

```diff
@@ -0,0 +1,289 @@
+    OperationResult,
+)
+from whylogs.core.metrics.serializers import serializer
+from whylogs.core.preprocessing import ListView, PreprocessedColumn
+from whylogs.core.proto import MetricComponentMessage, MetricMessage
+
+_SMALL = np.finfo(float).eps
+
+
+def _reciprocal(s: np.ndarray) -> np.ndarray:
+    """Return pseudoinverse of singular value vector"""
+    # should also zap if too small relative to s[0]
+    return np.array([1 / x if x > _SMALL else 0 for x in s])
+
+
+def _serialize_ndarray(a: np.ndarray) -> bytes:
+    bio = BytesIO()
+    np.save(bio, a, allow_pickle=False)
+    return bio.getvalue()
+
+
+def _deserialize_ndarray(a: bytes) -> np.ndarray:
+    bio = BytesIO(a)
+    return np.load(bio, allow_pickle=False)
+
+
+class VectorComponent(MetricComponent[np.ndarray]):
+    # mtype = np.ndarray
+    type_id = 101
+
+
+@serializer(type_id=101)
+def serialize(value: np.ndarray) -> MetricComponentMessage:
+    return MetricComponentMessage(serialized_bytes=_serialize_ndarray(value))
+
+
+@deserializer(type_id=101)
+def deserialize(msg: MetricComponentMessage) -> "VectorComponent":
+    return VectorComponent(value=_deserialize_ndarray(msg.serialized_bytes))
+
+
+"""
+    def to_protobuf(self) -> MetricComponentMessage:
+        return MetricComponentMessage(type_id=self.type_id, serialized_bytes=_serialize_ndarray(self.value))
+
+    @classmethod
+    def from_protobuf(cls, msg: MetricComponentMessage, registries: Optional[Registries] = None) -> "VectorComponent":
+        return VectorComponent(value=_deserialize_ndarray(msg.serialized_bytes))
+"""
+
+
+@dataclass(frozen=True)
+class SvdMetricConfig(MetricConfig):
+    k: int = 0
+    decay: float = 1.0
+
+
+# The SvdMetric classes just hold (and optionally update) the SVD
+# sketch. They can compute residuals from the current SVD approximation,
+# but they do not maintain any statistics about the distribution. See
+# the NlpMetric for tracking the residual distribution.
+
+
+@dataclass(frozen=True)
+class SvdMetric(Metric):
+    """
+    non-updating SVD metric
+    """
+
+    k: IntegralComponent  # SVD truncation  k > 0
+    decay: FractionalComponent  # 0 < decay <= 1  decay rate of old data
+    U: VectorComponent  # left singular vectors
+    S: VectorComponent  # singular values
+
+    @property
+    def namespace(self) -> str:
+        return "svd"
+
+    def residual(self, vector: np.ndarray) -> float:
+        """
+        Retruns the residual of the vector given the current approximate SVD:
+        residual = || U S S^{+} U' x - x || / || x ||  where x is the vector
+        """
+        U = self.U.value
+        S = self.S.value
+        residual = U.transpose() * vector
+        residual = _reciprocal(S) * residual
+        residual = S * residual
+        residual = U * residual
+        residual = residual - vector
+        residual = np.linalg.norm(residual) / np.linalg.norm(vector)
+        return residual
+
+    def merge(self, other: "SvdMetric") -> "SvdMetric":
```

Comment:
This doesn't appear to actually merge as it doesn't work with the other at all. Was it supposed to?

Reply:
Or does it only work with the over SVDMetric if it's an UpdatableSVGMetric?